File: C++/special_positions_in_a_binary_matrix.cpp
// no. 1582
#include <map>
using namespace std;
class Solution {
public:
  int numSpecial(vector<vector<int>> &mat) {
    int ans = 0;
    int n_rows = mat.size();
    int n_cols = mat[0].size();
    map<int, int> row_map;
    map<int, int> col_map;
    vector<int> one_row;
    vector<int> one_col;
    for (int i = 0; i < n_rows; i++) {
      for (int j = 0; j < n_cols; j++) {
        if (mat[i][j]) {
          one_row.push_back(i);
          one_col.push_back(j);
          row_map[i]++;
          col_map[j]++;
        }
      }
    }
    for (int i = 0; i < one_row.size(); i++) {
      if ((row_map[one_row[i]] == 1) & (col_map[one_col[i]] == 1)) {
        ans++;
      }
    }
    return ans;
  }
};

File: python/special_positions_in_a_binary_matrix.py
# no. 1582
class Solution:
    def numSpecial(self, mat: List[List[int]]) -> int:
        n_cols = len(mat[0])
        n_rows = len(mat)
        one_row, one_col = [], []
        row_dict, col_dict = {}, {}
        for i in range(n_rows):
            for j in range(n_cols):
                if mat[i][j]:
                    one_row.append(i)
                    one_col.append(j)
                    row_dict[i] = row_dict.get(i, 0) + 1
                    col_dict[j] = col_dict.get(j, 0) + 1
        ans = 0
        for i in range(len(one_row)):
            if row_dict[one_row[i]] == 1 and col_dict[one_col[i]] == 1:
                ans += 1
        return ans


